openem-ingestor/core/taskqueue.go

```diff
@@ -0,0 +1,163 @@
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/wailsapp/wails/v2/pkg/runtime"
+)
+
+type TaskQueue struct {
+	datasetSourceFolders sync.Map           // Datastructure to store all the upload requests
+	inputChannel         chan IngestionTask // Requests to upload data are put into this channel
+	resultChannel        chan TaskResult    // The result of the upload is put into this channel
+	AppContext           context.Context
+	Config               Config
+}
+
+type TransferMethods int
+
+const (
+	TransferS3 TransferMethods = iota + 1
+	TransferGlobus
+)
+
+type TransferOptions struct {
+	S3_endpoint string
+	S3_Bucket   string
+	Md5checksum bool
+}
+
+type IngestionTask struct {
+	// DatasetFolderId   uuid.UUID
+	DatasetFolder     DatasetFolder
+	ScicatUrl         string
+	ScicatAccessToken string
+	TransferMethod    TransferMethods
+	TransferOptions   TransferOptions
+	RequestContext    context.Context
+	Cancel            context.CancelFunc
+}
+
+type TaskResult struct {
+	Elapsed_seconds int
+	Dataset_PID     string
+	Error           error
+}
+
+func (w *TaskQueue) Startup() {
+
+	w.inputChannel = make(chan IngestionTask)
+	w.resultChannel = make(chan TaskResult)
+
+	// start multiple go routines/workers that will listen on the input channel
+	for worker := 1; worker <= w.Config.Misc.ConcurrencyLimit; worker++ {
+		go w.startWorker()
+	}
+
+}
+
+func (w *TaskQueue) CreateTask(folder DatasetFolder) error {
+
+	task := IngestionTask{
+		DatasetFolder:     folder,
+		ScicatUrl:         w.Config.Scicat.Host,
+		ScicatAccessToken: w.Config.Scicat.AccessToken,
+		TransferMethod:    TransferS3,
+		TransferOptions: TransferOptions{
+			S3_endpoint: w.Config.Transfer.S3.Endpoint,
+			S3_Bucket:   w.Config.Transfer.S3.Bucket,
+			Md5checksum: w.Config.Transfer.S3.Checksum,
+		},
+	}
+	_, found := w.datasetSourceFolders.Load(task.DatasetFolder.Id)
+	if found {
+		return errors.New("Key  exists")
+	}
+	w.datasetSourceFolders.Store(task.DatasetFolder.Id, task)
+	return nil
+}
+
+// Go routine that listens on the channel continously for upload requests and executes uploads.
+func (w *TaskQueue) startWorker() {
+	for ingestionTask := range w.inputChannel {
+		task_context, cancel := context.WithCancel(w.AppContext)
+		defer cancel()
+
+		ingestionTask.Cancel = cancel
+
+		result := w.IngestDataset(task_context, ingestionTask)
+		w.resultChannel <- result
+	}
+}
+
+func (w *TaskQueue) CancelTask(id uuid.UUID) {
+	value, found := w.datasetSourceFolders.Load(id)
+	if found {
+		f := value.(IngestionTask)
+		if f.Cancel != nil {
+			f.Cancel()
+		}
+		runtime.EventsEmit(w.AppContext, "upload-canceled", id)
+	}
+}
+
+func (w *TaskQueue) RemoveTask(id uuid.UUID) {
+	value, found := w.datasetSourceFolders.Load(id)
+	if found {
+		f := value.(IngestionTask)
+		if f.Cancel != nil {
+			f.Cancel()
+		}
+		w.datasetSourceFolders.Delete(id)
+		runtime.EventsEmit(w.AppContext, "folder-removed", id)
+	}
+}
+
+func (w *TaskQueue) ScheduleTask(id uuid.UUID) {
```

Comment:
"schedule" kind of implies that it is to be done in the future at a specified date. I feel like "queue" or "enqueue" would be the better wording here. Though with it already being part of a struct called "TaskQueue" that would be confusing as well... TaskQueue could also become "TaskRequests" since the requests are not in a queue yet. In general, I feel like the naming in general for TaskQueue and related functions could be improved